src/narrowdown.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (C) 2020 Veloman Yunkan
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * is provided AS IS, WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, and
+ * NON-INFRINGEMENT.  See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ *
+ */
+
+#ifndef ZIM_NARROWDOWN_H
+#define ZIM_NARROWDOWN_H
+
+#include "zim_types.h"
+#include "debug.h"
+
+#include <algorithm>
+#include <vector>
+
+namespace zim
+{
+
+class NarrowDown
+{
+public: // types
+  struct Range
+  {
+    article_index_type begin, end;
+  };
+
+public: // functions
+  void add(const std::string& key, article_index_type i)
+  {
+    ASSERT(entries.empty() || entries.back().keyPrefix < key, ==, true);
+    ASSERT(entries.empty() || entries.back().upperBound  < i,    ==, true);
+    entries.push_back({key, i});
+  }
+
+  Range getRange(const std::string& key) const
+  {
+    auto it = std::upper_bound(entries.begin(), entries.end(), key, LookupPred());
+    Range r;
+    if ( it == entries.end() )
+      --it;
+    r.end   = it->upperBound;
+    if ( it != entries.begin() )
+      --it;
+    r.begin = it->upperBound;
+    return r;
+  }
+
+private: // types
+  struct Entry
+  {
+    std::string keyPrefix;
```

Comment:
This is not a key prefix. This is "simply" a key. We are not looking a the value start with the prefix but simply compare the value with the key.

Reply:
I named the variable for the future. In later commits shortened pseudo-keys are stored in `zim::NarrowDown`. Yet, I renamed this data-member to `pseudoKey`.